applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.cpp

```diff
@@ -0,0 +1,332 @@
+
+        Matrix stress_design_variable_derivative;
+
+        if(mStressTreatment == StressTreatment::Mean)
+        {
+            rAdjointElem.Calculate(STRESS_DESIGN_DERIVATIVE_ON_GP, stress_design_variable_derivative, rProcessInfo);
+            this->ExtractMeanStressDerivative(stress_design_variable_derivative, rResponseGradient);
+        }
+        else if(mStressTreatment == StressTreatment::GaussPoint)
+        {
+            rAdjointElem.Calculate(STRESS_DESIGN_DERIVATIVE_ON_GP, stress_design_variable_derivative, rProcessInfo);
+            this->ExtractGaussPointStressDerivative(stress_design_variable_derivative, rResponseGradient);
+        }
+        else if(mStressTreatment == StressTreatment::Node)
+        {
+            rAdjointElem.Calculate(STRESS_DESIGN_DERIVATIVE_ON_NODE, stress_design_variable_derivative, rProcessInfo);
+            this->ExtractNodeStressDerivative(stress_design_variable_derivative, rResponseGradient);
+        }
+
+        KRATOS_ERROR_IF(rResponseGradient.size() != rDerivativesMatrix.size1())
+             << "Size of partial stress design variable derivative does not fit!" << std::endl;
+
+        rAdjointElem.SetValue(DESIGN_VARIABLE_NAME, "");
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::ExtractMeanStressDerivative(const Matrix& rStressDerivativesMatrix, Vector& rResponseGradient)
+    {
+        KRATOS_TRY;
+
+        const SizeType num_of_derivatives_per_stress = rStressDerivativesMatrix.size1();
+        const SizeType num_of_stress_positions = rStressDerivativesMatrix.size2();
+        double stress_derivative_value = 0.0;
+
+        if(rResponseGradient.size() != num_of_derivatives_per_stress)
+            rResponseGradient.resize(num_of_derivatives_per_stress, false);
+
+        for (IndexType deriv_it = 0 ; deriv_it < num_of_derivatives_per_stress; ++deriv_it)
+        {
+            for(IndexType stress_it = 0; stress_it < num_of_stress_positions; ++stress_it)
+                stress_derivative_value += rStressDerivativesMatrix(deriv_it, stress_it);
+
+            stress_derivative_value /= num_of_stress_positions;
+
+            rResponseGradient[deriv_it] = stress_derivative_value;
+            stress_derivative_value = 0.0;
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void AdjointLocalStressResponseFunction::ExtractNodeStressDerivative(const Matrix& rStressDerivativesMatrix, Vector& rResponseGradient)
+    {
+        KRATOS_TRY;
+
+        const SizeType num_of_derivatives_per_stress = rStressDerivativesMatrix.size1();
+        const SizeType num_of_stress_positions = rStressDerivativesMatrix.size2();
+
+        if(rResponseGradient.size() != num_of_derivatives_per_stress)
+            rResponseGradient.resize(num_of_derivatives_per_stress, false);
+
+        for (IndexType deriv_it = 0 ; deriv_it < num_of_derivatives_per_stress; ++deriv_it)
+        {
+            if(num_of_stress_positions >= mIdOfLocation)
```

Comment:
Don't put the if statement in the loop. Just do `KRATOS_ERROR_IF_NOT(num_of_stress_positions >= mIdOfLocation)` before it.